File: mobile/src/services/deviceClassifier.ts
import type { DeviceType, Port } from '@shared/src/types/device';
import { DEVICE_TYPE_PATTERNS } from '@shared/src/constants';

export function classifyDevice(
  mac: string,
  vendor?: string,
  hostname?: string,
  openPorts?: Port[]
): DeviceType {
  const searchText = [
    vendor,
    hostname,
    ...(openPorts?.map(p => p.service) || []),
  ]
    .filter(Boolean)
    .join(' ')
    .toLowerCase();

  for (const [deviceType, patterns] of Object.entries(DEVICE_TYPE_PATTERNS)) {
    for (const pattern of patterns) {
      if (searchText.includes(pattern.toLowerCase())) {
        return deviceType as DeviceType;
      }
    }
  }

  if (openPorts) {
    if (openPorts.some(p => p.number === 554 || p.number === 8554)) {
      return 'camera';
    }
    if (openPorts.some(p => p.number === 80 || p.number === 443 || p.number === 8080)) {
      if (vendor && isKnownCameraVendor(vendor)) {
        return 'camera';
      }
    }
  }

  if (vendor) {
    if (isKnownRouterVendor(vendor)) {
      return 'router';
    }
    if (isKnownCameraVendor(vendor)) {
      return 'camera';
    }
  }

  if (hostname) {
    const hostnameLower = hostname.toLowerCase();
    if (hostnameLower.includes('router') || hostnameLower.includes('gateway')) {
      return 'router';
    }
  }

  return 'unknown';
}

function isKnownCameraVendor(vendor: string): boolean {
  const cameraVendors = [
    'hikvision',
    'dahua',
    'foscam',
    'amcrest',
    'reolink',
    'axis',
    'bosch',
    'panasonic',
    'honeywell',
    'nest',
    'ring',
    'arlo',
    'logitech',
    'netatmo',
    ' Canary',
    'dropcam',
    'logitech',
    'yi',
    'xiaomi',
    'ezviz',
    'tplink',
    'tapo',
  ];
  
  const vendorLower = vendor.toLowerCase();
  return cameraVendors.some(v => vendorLower.includes(v));
}

function isKnownRouterVendor(vendor: string): boolean {
  const routerVendors = [
    'cisco',
    'linksys',
    'netgear',
    'tp-link',
    'd-link',
    'asus',
    'ubiquiti',
    'mikrotik',
    'arris',
    'netcomm',
    'technicolor',
    'sagemcom',
    'thomson',
    'huawei',
    'zyxel',
    'buffalo',
    'trendnet',
    'linksys',
    'belkin',
    'tenda',
  ];
  
  const vendorLower = vendor.toLowerCase();
  return routerVendors.some(v => vendorLower.includes(v));
}

File: mobile/modules/network-scanner/android/src/main/java/com/camdetect/app/DeviceClassifier.java
package com.camdetect.app;

import java.util.Arrays;
import java.util.List;

public class DeviceClassifier {
    
    private static final List<String> CAMERA_VENDORS = Arrays.asList(
        "hikvision", "dahua", "foscam", "amcrest", "reolink", "axis",
        "bosch", "honeywell", "nest", "ring", "arlo", "logitech",
        "netatmo", "canary", "dropcam", "yi", "xiaomi", "ezviz", "tapo"
    );
    
    private static final List<String> ROUTER_VENDORS = Arrays.asList(
        "cisco", "linksys", "netgear", "tp-link", "d-link", "asus",
        "ubiquiti", "mikrotik", "arris", "technicolor", "sagemcom",
        "huawei", "zyxel", "buffalo", "trendnet", "belkin", "tenda"
    );
    
    public static String classify(String vendor, String hostname, List<PortScanner.PortInfo> ports) {
        String searchText = (vendor != null ? vendor : "") + " " + (hostname != null ? hostname : "");
        searchText = searchText.toLowerCase();
        
        if (vendor != null) {
            String vendorLower = vendor.toLowerCase();
            for (String cv : CAMERA_VENDORS) {
                if (vendorLower.contains(cv)) return "camera";
            }
            for (String rv : ROUTER_VENDORS) {
                if (vendorLower.contains(rv)) return "router";
            }
        }
        
        if (ports != null) {
            for (PortScanner.PortInfo port : ports) {
                if (port.number == 554 || port.number == 8554) {
                    return "camera";
                }
            }
        }
        
        if (hostname != null) {
            String hostnameLower = hostname.toLowerCase();
            if (hostnameLower.contains("router") || hostnameLower.contains("gateway")) {
                return "router";
            }
        }
        
        return "unknown";
    }
    
    public static boolean isKnownCameraVendor(String vendor) {
        if (vendor == null) return false;
        String vendorLower = vendor.toLowerCase();
        for (String cv : CAMERA_VENDORS) {
            if (vendorLower.contains(cv)) return true;
        }
        return false;
    }
}
